### PR TITLE
Fix #4, #70, #71, #86, #87: Introduce data module

### LINKED ISSUE #4: Project introduction
The initial Android Studio project structure should be introduced with each of the 5 modules described in the technical design document, with some interactions between them. These modules comprise of:
- [x] App module: this contains the frontend Android code & UIs
- [x] Data-specification module: this contains protos to be used in other modules as the domain model
- [x] Domain module: this contains business logic to be consumed by the app module
- [ ] Data source module: this contains code that interacts with the disk and Oppia's backends to provide data needed by the domain module; to be consumed by the domain module
- [x] Utility module: this contains reusable UI utility components, and the data source mechanism that's used to share data between the data source and domain modules

### PR DESCRIPTION
**Pull request by BenHenning**
First, this introduces a fully functioning data module. It also introduces a lot of constituent pieces in order to make that data module work:
- Dagger introduction
- PersistentCacheStore
- DataProvider framework

Functionally, this results in the Oppia app now properly showing "Welcome to Oppia!" on initial app open, and "Welcome back to Oppia!" upon returning to the app. Both of these states are also properly retained across configuration changes due to Dagger not leading to activities trying to recreate their dependent controllers.

From a testing side, this added new HomeActivity tests which rely on sleeping rather than idling resource. Long-term once #59 is resolved, we'll be able to consolidate the ``runBlockingTest`` from Robolectric and idling resource from Espresso into a consistent blocker that waits for real background-executed coroutines to fully resolve before proceeding to continue in tests. In the meantime, this PR updates MainActivity's Espresso behavior to use the AndroidX test orchestrator and properly reset app state across test runs. This is essential to ensure the on-disk cache is removed and reinitialized in each test run.

The Dagger introduction here is loosely based on a combination of https://proandroiddev.com/dagger-2-on-android-the-simple-way-f706a2c597e9 and a custom version of Dagger-Android (https://dagger.dev/android.html). We should eventually document and land on a final Dagger approach, but I like this implementation since it's simpler than Dagger-Android, and maximizes injectability. It introduces application, activity, and fragment scopes to provide more granular control over when dependencies should be recreated. It also introduces the concept of application, activity, and fragment controllers to perform actual logic, keeping activities and fragments essentially thin classes that delegate responsibility to their controllers. View models are also injectable, but I'm not sure whether we want to keep this long-term. In order to hook things up correctly, activities and fragments should extend special injectable base activity/fragment classes. Although this additional inheritance isn't great, it's reasonable if we don't add anything to these base classes. Long-term we might be able to replace this with code generation.

This PR introduces a fully-functioning DataProvider framework. This needs to be fully designed yet, as the current implementation will have value and error channel elision issues when converted to LiveData. However, this is a proposed start that lets us perform complex data pipeline transformations that are guaranteed to execute on background threads, and provides complete Kotlin coroutine interop. The conversion to LiveData also happens at the end, ensuring there's one LiveData per UI subscription (which I noticed was mentioned as a possible best practice somewhere, but I can't recall the source at the moment).

Finally, this PR also introduces an on-disk cache that's also a DataProvider. This cache stores protos and is expected to be a viable long-term replacement to SharedPreferences. I ran into a few issues with doing this in a clean way: in the test activity, we're setting that the user opened the app early on. However, we don't want this to change the on-disk file before it's read, and we don't want the change to notify subscribers. This resulted in added complexity:
- An InMemoryCache was introduced to synchronize read/write operations
- Write notifications can be silenced in the cache store
- The in-memory cache of the store can be 'primed' to ensure that writes which occur before subscriptions trigger a read do not get picked up

This PR introduces reasonable tests for the components being covered here, but they are far from comprehensive. Observable behaviors in HomeActivity are at least tested, where all of the infrastructure introduced here is thoroughly integrated. Over time, these tests should be refined as the infrastructure itself matures. There's also some duplication between the tests that will be easier to clean up once #59 is resolved.

### CODE REVIEW

File: utility/src/main/java/org/oppia/util/data/AsyncDataSubscriptionManager.kt [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+import org.oppia.util.threading.ConcurrentQueueMap
+import org.oppia.util.threading.dequeue
+import org.oppia.util.threading.enqueue
+import org.oppia.util.threading.getQueue
+import javax.inject.Inject
+import javax.inject.Singleton
+
+internal typealias ObserveAsyncChange = suspend () -> Unit
+
+// TODO(BenHenning): Move this to the Dagger graph rather than making it a singleton.
```
Comment (seanlip) [inline utility/src/main/java/org/oppia/util/data/AsyncDataSubscriptionManager.kt]: Here and elsewhere (e.g. HomeActivityTest.kt): please link TODOs to issue numbers.

File: utility/src/main/java/org/oppia/util/data/AsyncDataSubscriptionManager.kt [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+class AsyncDataSubscriptionManager @Inject constructor() {
+  private val subscriptionMap = ConcurrentQueueMap<Any, ObserveAsyncChange>()
+  private val associatedIds = ConcurrentQueueMap<Any, Any>()
+
+  /** Subscribes the specified callback function to the specified [DataProvider] ID. */
+  internal fun subscribe(id: Any, observeChange: ObserveAsyncChange) {
+    subscriptionMap.enqueue(id, observeChange)
+  }
+
+  /** Ubsubscribes the specified callback function from the specified [DataProvider] ID. */
```
Comment (seanlip) [inline utility/src/main/java/org/oppia/util/data/AsyncDataSubscriptionManager.kt]: Nit: Unsubscribes

File: utility/src/main/java/org/oppia/util/data/AsyncDataSubscriptionManager.kt (hunk 1 above)
Comment (BenHenning) [reply utility/src/main/java/org/oppia/util/data/AsyncDataSubscriptionManager.kt]: Your comment came before my commit to actually do this. :) Done.

File: utility/src/main/java/org/oppia/util/data/AsyncDataSubscriptionManager.kt (hunk 2 above)
Comment (BenHenning) [reply utility/src/main/java/org/oppia/util/data/AsyncDataSubscriptionManager.kt]: Done.

**Comment by BenHenning** [comment]
This is now ready to review. PTAL!

File: utility/src/main/java/org/oppia/util/threading/ConcurrentCollections.kt [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+ * successful.
+ */
+fun <K, V> ConcurrentQueueMap<K, V>.dequeue(key: K, value: V): Boolean {
+  // See warning in enqueue() for scenarios when this deletion may fail.
+  return getQueue(key).remove(value)
+}
+
+/** Returns the [ConcurrentLinkedQueue] corresponding to the specified key in a thread-safe way. */
+internal fun <K, V> ConcurrentQueueMap<K, V>.getQueue(key: K): ConcurrentLinkedQueue<V> {
+  // NB: This is a pre-24 compatible alternative to computeIfAbsent. See: https://stackoverflow.com/a/40665232.
```
Comment (rt4914) [inline utility/src/main/java/org/oppia/util/threading/ConcurrentCollections.kt:29]: @BenHenning comments are not according to Javadocs. Check here and everywhere else too.

**Review by rt4914 — COMMENTED** [review]
Update comments as per JavaDocs.
Other than comments everything is good.
@BenHenning Maybe we can merge this for now and updates comments later on?

**Review by rt4914 — APPROVED** [review]
LGTM

**Comment by BenHenning** [comment]
Thanks @rt4914. I'll follow up on your comment separately if needed.

**Pull request merged**